### ISSUE TITLE
The training looks like not converging

### ISSUE THREAD

**Issue by shijin1984** [shijin1984/learn_rnn#1]
The loss and error rate keeps high, possibly the gradient is incorrect?

**Comment by shijin1984** [comment]
Used an example of LSTM, and the accuracy is as high as **98%**:
- Two stacks of LSTM, the output dimensions are both 128
- A 32-node hidden FC layer.
- The output layer with softmax.

To compare with my simple model, I reduce the network until it is:
- LSTM, whose output is 10
- Feed the output directly into softmax, without any hidden layers.

In other word, the only difference is LSTM v.s. Elman.
But still the LSTM can achieve accuracy of **87%**.

Details:
```
Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #
=================================================================
lstm (LSTM)                  (None, 10)                1560
_________________________________________________________________
softmax (Softmax)            (None, 10)                0
=================================================================
Total params: 1,560
Trainable params: 1,560
Non-trainable params: 0
_________________________________________________________________
Epoch 1/3
1875/1875 [==============================] - 16s 9ms/step - loss: 1.1188 - accuracy: 0.6095 - val_loss: 0.5823 - val_accuracy: 0.8147
Epoch 2/3
1875/1875 [==============================] - 16s 9ms/step - loss: 0.5293 - accuracy: 0.8358 - val_loss: 0.4453 - val_accuracy: 0.8626
Epoch 3/3
1875/1875 [==============================] - 16s 9ms/step - loss: 0.4161 - accuracy: 0.8731 - val_loss: 0.3965 - val_accuracy: 0.8767
```

**Comment by shijin1984** [comment]
Implemented the Elman cell in keras, and it can achieve accuracy of **64%** with the SGD optimizer.
Even with the builtin SimpleRNNCell, which is less powerful than Elman, the accuracy is **46%**.

So my implementation must be problematic.

**Comment by shijin1984** [comment]
A like cause is the initialization.

The optimizing progress stuck with changes in parameter initializer:
```
diff --git a/keras/keras_rnn.py b/keras/keras_rnn.py
index a33fa8c..8062e6d 100644
--- a/keras/keras_rnn.py
+++ b/keras/keras_rnn.py
@@ -1,6 +1,7 @@
 import tensorflow as  tf
 from tensorflow.keras import backend as K
 from tensorflow.keras import activations
+from tensorflow.keras.initializers import RandomUniform
 from tensorflow.keras.models import Sequential
 from tensorflow.keras.layers import Layer, RNN, Softmax

@@ -15,15 +16,20 @@ class ElmanCell(Layer):
     # h(t) = activation(
     #     Wh x(t) + Uh h(t-1) + bh)
     self.Wh = self.add_weight(shape=(input_shape[-1], self.state_size),
+                              initializer=RandomUniform(0, 1),
                               name='Wh')
     self.Uh = self.add_weight(shape=(self.state_size, self.state_size),
+                              initializer=RandomUniform(0, 1),
                               name='Uh')
     self.bh = self.add_weight(shape=(self.state_size,),
+                              initializer=RandomUniform(-0.5, 0.5),
                               name='bh')
     # y(t) = activate(Wy h(t) + by)
     self.Wy = self.add_weight(shape=(self.state_size, self.units),
+                              initializer=RandomUniform(0, 1),
                               name='Wy')
     self.by = self.add_weight(shape=(self.units,),
+                              initializer=RandomUniform(-0.5, 0.5),
                               name='by')
     self.built = True
```

The details:

```
Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #
=================================================================
rnn (RNN)                    (None, 10)                500
_________________________________________________________________
softmax (Softmax)            (None, 10)                0
=================================================================
Total params: 500
Trainable params: 500
Non-trainable params: 0
_________________________________________________________________
Epoch 1/10
1875/1875 [==============================] - 15s 8ms/step - loss: 1005449428074496.0000 - accuracy: 0.1035 - val_loss: 2.3015 - val_accuracy: 0.1135
Epoch 2/10
1875/1875 [==============================] - 15s 8ms/step - loss: 2.3018 - accuracy: 0.1124 - val_loss: 2.3016 - val_accuracy: 0.1135
Epoch 3/10
1875/1875 [==============================] - 15s 8ms/step - loss: 2.3018 - accuracy: 0.1124 - val_loss: 2.3016 - val_accuracy: 0.1135
Epoch 4/10
1875/1875 [==============================] - 14s 8ms/step - loss: 2.3018 - accuracy: 0.1124 - val_loss: 2.3016 - val_accuracy: 0.1135
Epoch 5/10
1875/1875 [==============================] - 15s 8ms/step - loss: 2.3018 - accuracy: 0.1124 - val_loss: 2.3016 - val_accuracy: 0.1135
Epoch 6/10
1875/1875 [==============================] - 15s 8ms/step - loss: 2.3018 - accuracy: 0.1124 - val_loss: 2.3016 - val_accuracy: 0.1135
Epoch 7/10
1875/1875 [==============================] - 15s 8ms/step - loss: 2.3018 - accuracy: 0.1124 - val_loss: 2.3016 - val_accuracy: 0.1135
Epoch 8/10
1875/1875 [==============================] - 15s 8ms/step - loss: 2.3018 - accuracy: 0.1124 - val_loss: 2.3016 - val_accuracy: 0.1135
Epoch 9/10
1875/1875 [==============================] - 15s 8ms/step - loss: 2.3018 - accuracy: 0.1124 - val_loss: 2.3016 - val_accuracy: 0.1135
Epoch 10/10
1875/1875 [==============================] - 15s 8ms/step - loss: 2.3018 - accuracy: 0.1124 - val_loss: 2.3016 - val_accuracy: 0.1135
```

**Comment by shijin1984** [comment]
To double check, I explicitly specify the initializers as the supposed default value:

```
index a33fa8c..00b3dff 100644
--- a/keras/keras_rnn.py
+++ b/keras/keras_rnn.py
@@ -1,6 +1,7 @@
 import tensorflow as  tf
 from tensorflow.keras import backend as K
 from tensorflow.keras import activations
+from tensorflow.keras.initializers import RandomUniform
 from tensorflow.keras.models import Sequential
 from tensorflow.keras.layers import Layer, RNN, Softmax

@@ -15,15 +16,20 @@ class ElmanCell(Layer):
     # h(t) = activation(
     #     Wh x(t) + Uh h(t-1) + bh)
     self.Wh = self.add_weight(shape=(input_shape[-1], self.state_size),
+                              initializer='glorot_uniform',
                               name='Wh')
     self.Uh = self.add_weight(shape=(self.state_size, self.state_size),
+                              initializer='glorot_uniform',
                               name='Uh')
     self.bh = self.add_weight(shape=(self.state_size,),
+                              initializer='zeros',
                               name='bh')
     # y(t) = activate(Wy h(t) + by)
     self.Wy = self.add_weight(shape=(self.state_size, self.units),
+                              initializer='glorot_uniform',
                               name='Wy')
     self.by = self.add_weight(shape=(self.units,),
+                              initializer='zeros',
                               name='by')
     self.built = True
```

It is not very stable, but at least it usually works!

The initializer is critical.

**Comment by shijin1984** [comment]
The LSTM works finally!

For the Elman network, it is also a little bit unstable on Keras. I think it is because Keras has some features like regularization can improve the training process, and mine does not work.
